apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/ConfigsExportService.java

```diff
@@ -0,0 +1,181 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Service;
+
+@Service
+public class ConfigsExportService {
+
+  private final static Logger logger = LoggerFactory.getLogger(ConfigsExportService.class);
+
+  private final AppService appService;
+
+  private final ClusterService clusterService;
+
+  private final NamespaceService namespaceService;
+
+  private final PortalSettings portalSettings;
+
+  private final PermissionValidator permissionValidator;
+
+  public ConfigsExportService(
+      AppService appService,
+      ClusterService clusterService,
+      final @Lazy NamespaceService namespaceService,
+      PortalSettings portalSettings,
+      PermissionValidator permissionValidator) {
+    this.appService = appService;
+    this.clusterService = clusterService;
+    this.namespaceService = namespaceService;
+    this.portalSettings = portalSettings;
+    this.permissionValidator = permissionValidator;
+  }
+
+  /**
+   * write multiple namespace to a zip.
+   * use {@link Stream#reduce(Object, BiFunction, BinaryOperator)} to forbid concurrent write.
+   * @param configBOStream namespace's stream
+   * @param outputStream receive zip file output stream
+   * @throws IOException
+   */
+  public static void writeAsZipOutputStream(Stream<ConfigBO> configBOStream, OutputStream outputStream)
+      throws IOException {
+    try (final ZipOutputStream zipOutputStream = new ZipOutputStream(outputStream)) {
+      final Consumer<ConfigBO> configBOConsumer = configBO -> {
+        try {
+          // TODO, Stream.reduce will cause some problems. Is There other way to speed up the downloading?
+          synchronized (zipOutputStream) {
+            write2ZipOutputStream(zipOutputStream, configBO);
+          }
+        } catch (IOException e) {
+          logger.error("Write error. {}", configBO);
+          throw new IllegalStateException(e);
+        }
+      };
+      configBOStream.forEach(configBOConsumer);
+    }
+  }
+
+  /**
+   * write {@link ConfigBO} as file to {@link ZipOutputStream}.
+   * Watch out the concurrent problem!
+   * zip output stream is same like cannot write concurrently!
+   * the name of file is determined by {@link ConfigFileUtils#toFilename(String, String, String, ConfigFileFormat)}.
+   * the path of file is determined by {@link ConfigFileUtils#toFilePath(String, String, Env, String)}.
+   * @param zipOutputStream zip file output stream
+   * @param configBO a namespace represent
+   * @return zip file output stream same as parameter zipOutputStream
+   */
+  public static ZipOutputStream write2ZipOutputStream(
+      final ZipOutputStream zipOutputStream,
+      final ConfigBO configBO
+  ) throws IOException {
+    final Env env = configBO.getEnv();
```

Comment:
what is the usage of `final` here?

Reply:
Just a trick to prevent me from writing wrong code. (Like wrong assign operation).